src/runner.c

```diff
@@ -0,0 +1,134 @@
+    CFStringRef path = (CFStringRef)CFArrayGetValueAtIndex((CFArrayRef)eventPaths, idx);
+    event->pathlen = CFStringGetLength(path);
+    strncpy(event->path, CFStringGetCStringPtr(path, kCFStringEncodingUTF8), PATH_MAX);
+    event->id = eventIds[idx];
+    event->flags = eventFlags[idx];
+    event->next = NULL;
+    if (instance->head) {
+      ((fse_event_t*)instance->head)->next = event;
+      instance->head = event;
+    } else {
+      instance->head = event;
+      instance->tail = event;
+    }
+  }
+  uv_async_send(&instance->async);
+  uv_mutex_unlock(&instance->mutex);
+}
+
+void eventCallback(uv_async_t *async) {
+  fse_t *instance = (fse_t *) async->data;
+  napi_handle_scope scope;
+  napi_value this, callback, argv[3], result;
+  fse_event_t *event;
+  uv_mutex_lock(&instance->mutex);
+  assert(napi_open_handle_scope(instance->env, &scope) == napi_ok);
+  assert(napi_get_null(instance->env, &this) == napi_ok);
+  assert(napi_get_reference_value(instance->env, instance->callback, &callback) == napi_ok);
+  instance->head = NULL;
+  while (instance->tail) {
+    event = (fse_event_t*)instance->tail;
+    instance->tail = event->next;
+    assert(napi_create_string_utf8(instance->env, event->path, event->pathlen, &argv[0]) == napi_ok);
+    assert(napi_create_uint32(instance->env, event->flags, &argv[1]) == napi_ok);
+    assert(napi_create_int64(instance->env, event->id, &argv[2]) == napi_ok);
+    napi_call_function(instance->env, this, callback, 3, argv, &result);
+    free(event);
+  }
+  assert(napi_close_handle_scope(instance->env, scope) == napi_ok);
+  uv_mutex_unlock(&instance->mutex);
+}
+
+void execute(void *data) {
+  fse_t *instance = (fse_t*)data;
+
+  instance->loop = CFRunLoopGetCurrent();
+
+  FSEventStreamContext context = { 0, data, NULL, NULL, NULL };
+  CFStringRef dirs[] = { CFStringCreateWithCString(NULL, instance->path, kCFStringEncodingUTF8) };
+  FSEventStreamRef stream = FSEventStreamCreate(NULL, &handleEvents, &context, CFArrayCreate(NULL, (const void **)&dirs, 1, NULL), kFSEventStreamEventIdSinceNow, (CFAbsoluteTime) 0.1, kFSEventStreamCreateFlagNone | kFSEventStreamCreateFlagWatchRoot | kFSEventStreamCreateFlagFileEvents | kFSEventStreamCreateFlagUseCFTypes);
+  FSEventStreamScheduleWithRunLoop(stream, instance->loop, kCFRunLoopDefaultMode);
+  FSEventStreamStart(stream);
+
+  CFRunLoopRun();
+
+  FSEventStreamStop(stream);
+  FSEventStreamUnscheduleFromRunLoop(stream, instance->loop, kCFRunLoopDefaultMode);
+  FSEventStreamInvalidate(stream);
+  FSEventStreamRelease(stream);
+
+  instance->loop = NULL;
+}
+
+void cleanup(napi_env env, void* data, void* ignored) {
+  fse_t *instance = (fse_t*)data;
+  stop(instance);
+  assert(napi_delete_reference(instance->env, instance->callback) == napi_ok);
+  free(instance);
+}
+
+void stop(fse_t *instance) {
+  if (!instance->loop) return;
+  unsigned int count = 0;
+  CFRunLoopStop(instance->loop);
+  uv_thread_join(&instance->thread);
```

Comment:
Check the return value. An error is probably abort-worthy; you could turn it into a CHECK.